File: Fortran/ImpulseResponses.f90
SUBROUTINE ImpulseResponses

USE Parameters
USE Globals
USE Procedures

IMPLICIT NONE
INTEGER		:: it,it1
CHARACTER	:: IRFDir*20

!allocate arrays
CALL AllocateArrays

stickytransition = .false.

!set up deltatransvec
CALL PowerSpacedGrid (Ttransition,deltatransparam,deltatransmin,deltatransmax,deltatransvec)

nendtrans = min(50,Ttransition)
cumdeltatrans(1) = deltatransvec(1)
DO it = 2,Ttransition
	cumdeltatrans(it) = cumdeltatrans(it-1) + deltatransvec(it)
END DO
OPEN(3, FILE = trim(OutputDir) // 'deltatransvec.txt', STATUS = 'replace'); CALL WriteMatrix(3,Ttransition,1,deltatransvec)



!Monetary policy shock
IF(IncludeMonetaryShock==1) THEN
	IF(Display>=1) write(*,*)'Solving for monetary policy shock IRF'	
	irfpointer => irfstruct(0)

	equmTRANS(:) = equmINITSS		
	equmTRANS(1)%mpshock = equmINITSS%mpshock + MonetaryShockSize
	DO it = 2,Ttransition-nendtrans
		equmTRANS(it)%mpshock =equmINITSS%mpshock *(1.0-MonetaryShockPers**deltatransvec(it-1)) + equmTRANS(it-1)%mpshock * (MonetaryShockPers**deltatransvec(it-1))

	END DO	
	equmTRANS(Ttransition-nendtrans+1:Ttransition)%mpshock = equmINITSS%mpshock

	IRFDir = "Monetary"
	CALL IRFSequence(IRFDir)

END IF

!Forward Guidance shock
IF(IncludeForwardGuideShock==1) THEN
	forwardguide = .true.
	IF(Display>=1) write(*,*)'Solving for forward guidance shock IRF'	
	irfpointer => irfstruct(0)

	equmTRANS(:) = equmINITSS
	it1 = MINLOC(cumdeltatrans, 1, MASK = cumdeltatrans>=ForwardGuideShockQtrs)
	equmTRANS(1:it1-1)%mpshock = equmINITSS%mpshock
	equmTRANS(it1)%mpshock = equmINITSS%mpshock + ForwardGuideShockSize
	DO it = it1+1,Ttransition-nendtrans
		equmTRANS(it)%mpshock =equmINITSS%mpshock *(1.0-MonetaryShockPers**deltatransvec(it-1)) + equmTRANS(it-1)%mpshock * (MonetaryShockPers**deltatransvec(it-1))

	END DO	
	equmTRANS(Ttransition-nendtrans+1:Ttransition)%mpshock = equmINITSS%mpshock

	IRFDir = "ForwardGuide"
	CALL IRFSequence(IRFDir)
	forwardguide = .false.

END IF


END SUBROUTINE ImpulseResponses

File: mycode/ImpulseResponses.m
global deltatransvec

stickytransition = false;

% set up deltatransvec
deltatransvec = PowerSpacedGrid(Ttransition,deltatransparam,deltatransmin,deltatransmax);
cumdeltatrans = cumsum(deltatransvec);

equmTRANS = struct;
equmTRANS.borrwedge = equmINITSS.borrwedge;
equmTRANS.fundlev = equmINITSS.fundlev;
equmTRANS.elast = equmINITSS.elast;
equmTRANS.tfp = equmINITSS.tfp;
equmTRANS.caputil = equmINITSS.caputil;
equmTRANS.labtax = equmINITSS.labtax;

nendtrans = min(50,Ttransition);
irfstruct = struct;

% Monetary policy shock
if IncludeMonetaryShock
    if Display; fprintf('Solving for monetary policy shock IRF\n'); end
    
    equmTRANS.mpshock = equmINITSS.mpshock + MonetaryShockSize * MonetaryShockPers.^[0 cumdeltatrans(1:Ttransition-1)];
    equmTRANS.mpshock(end-nendtrans+1:end) = equmINITSS.mpshock;

    IRFDir = 'Monetary';
    IRFSequence
end

% Forward Guidance shock
if IncludeForwardGuideShock
    if Display; fprintf('Solving for forward guidance shock IRF\n'); end
    
    % forward guidance time
    itfg = find(cumdeltatrans>=ForwardGuideShockQtrs,1);
    equmTRANS.mpshock = equmINITSS.mpshock + ForwardGuideShockSize * MonetaryShockPers.^[zeros(1,itfg) cumsum(deltatransvec(itfg:Ttransition-1))];
    equmTRANS.mpshock(end-nendtrans+1:end) = equmINITSS.mpshock;

    IRFDir = 'ForwardGuide';
    IRFSequence
end